keyboards/cipulot/ec_pro2/ec_switch_matrix.c

```diff
@@ -0,0 +1,182 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ec_switch_matrix.h"
+#include "analog.h"
+#include "atomic_util.h"
+#include "print.h"
+
+/* Pin and port array */
+const uint32_t row_pins[]     = MATRIX_ROW_PINS;
+const uint8_t  col_channels[] = MATRIX_COL_CHANNELS;
+const uint32_t mux_sel_pins[] = MUX_SEL_PINS;
+
+static ecsm_config_t config;
+static uint16_t      ecsm_sw_value[MATRIX_ROWS][MATRIX_COLS];
+
+static adc_mux adcMux;
+
+static inline void discharge_capacitor(void) {
+    ATOMIC_BLOCK_FORCEON {
+        writePinLow(DISCHARGE_PIN);
+    }
+}
+static inline void charge_capacitor(uint8_t row) {
+    ATOMIC_BLOCK_FORCEON {
+        writePinHigh(DISCHARGE_PIN);
+        writePinHigh(row_pins[row]);
+    }
+}
+
+static inline void init_mux_sel(void) {
+    for (int idx = 0; idx < 3; idx++) {
+        setPinOutput(mux_sel_pins[idx]);
+    }
+}
+
+static inline void select_mux(uint8_t col) {
+    uint8_t ch = col_channels[col];
+    writePin(mux_sel_pins[0], ch & 1);
+    writePin(mux_sel_pins[1], ch & 2);
+    writePin(mux_sel_pins[2], ch & 4);
+}
+
+static inline void init_row(void) {
+    for (int idx = 0; idx < MATRIX_ROWS; idx++) {
+        setPinOutput(row_pins[idx]);
+        writePinLow(row_pins[idx]);
+    }
+}
+
+/* Initialize the peripherals pins */
+int ecsm_init(ecsm_config_t const* const ecsm_config) {
+    // Initialize config
+    config = *ecsm_config;
+
+    palSetLineMode(ANALOG_PORT, PAL_MODE_INPUT_ANALOG);
+    adcMux = pinToMux(ANALOG_PORT);
+
+    // Initialize discharge pin as discharge mode
+    writePinLow(DISCHARGE_PIN);
+    setPinOutputOpenDrain(DISCHARGE_PIN);
+
+    // Initialize drive lines
+    init_row();
+
+    // Initialize multiplexer select pin
+    init_mux_sel();
+
+    // Enable AMUX
+    setPinOutput(APLEX_EN_PIN_0);
+    writePinLow(APLEX_EN_PIN_0);
+    setPinOutput(APLEX_EN_PIN_1);
+    writePinLow(APLEX_EN_PIN_1);
+
+    return 0;
+}
+
+int ecsm_update(ecsm_config_t const* const ecsm_config) {
+    // Save config
+    config = *ecsm_config;
+    return 0;
+}
+
+// Read the capacitive sensor value
+uint16_t ecsm_readkey_raw(uint8_t channel, uint8_t row, uint8_t col) {
+    uint16_t sw_value = 0;
+
+    // Select the multiplexer
+    if (channel == 0) {
+        writePinHigh(APLEX_EN_PIN_0);
+        select_mux(col);
+        writePinLow(APLEX_EN_PIN_0);
+    } else {
+        writePinHigh(APLEX_EN_PIN_1);
+        select_mux(col);
+        writePinLow(APLEX_EN_PIN_1);
+    }
+
+    // Set strobe pins to low state
+    writePinLow(row_pins[row]);
+
+    // Set the row pin to high state and have capacitor charge
+    charge_capacitor(row);
+
+    ATOMIC_BLOCK_FORCEON {
+        sw_value = adc_read(adcMux);
+    }
```

Comment:
Technically this is wrong, because `ATOMIC_BLOCK_FORCEON` is `chSysLock(); ...; chSysUnlock();`, and `adcConvert()` has its own `osalSysLock()`/`osalSysUnlock()` pair inside.  This code will panic if `CH_DBG_SYSTEM_STATE_CHECK` is enabled; it just happens to work without debug checks, because the duplicate lock and unlock calls are ignored.

Doing this properly may require either some new analog API, or using the ChibiOS ADC driver directly (`adcStartConversionI()` can be called from the S-Locked state, but waiting for the conversion completion would require working interrupts, so it can't be done without leaving the S-Locked state temporarily — either explicitly, or through a thread suspend, which is what `adcConvert()` does internally).

Also I suppose that what you want here is to avoid interrupts messing up timings — but in that case the atomic block should also cover at least `charge_capacitor(row);` instead of having a separate atomic block inside that function (otherwise an interrupt can be served between those atomic blocks).

Reply:
I see so the options are to wrap the `charge_capacitor(row)`, `sw_value = adc_read(adcMux)` and `discharge_capacitor()` in a single `ATOMIC_BLOCK_FORCEON` and remove the ones in the functions or remove it all together. 

I tested the first option and the board behaves as expected. If that's an ok solution for you @sigprof I'll push a commit for that.

The change basically looks like this:
```c
ATOMIC_BLOCK_FORCEON {
        // Set the row pin to high state and have capacitor charge
        charge_capacitor(row);

        ATOMIC_BLOCK_FORCEON {
            sw_value = adc_read(adcMux);
        }
        // Discharge peak hold capacitor
        discharge_capacitor();
    }
```
I'll wait for your confirmation to push this.

Reply:
Wrapping `discharge_capacitor()` into the atomic block is probably not needed — `adc_read()` will reenable interrupts anyway while it's waiting for the ADC conversion to finish, and some other interrupt can happen during that time.  Does an extra delay before `discharge_capacitor()` matter here?

As for making the code actually correct (and not just “works with debug checks disabled”), I'm not sure what could be done here.  It could be possible to add something like `adc_read_atomic()` that is safe to call inside an atomic block (by replacing the `adcConvert()` call in the existing implementation with a copy of its internal implementation with lock/unlock removed); but it won't be really atomic, because waiting for the conversion completion needs to be done with interrupts enabled.  Also there would be some issues with the ADC initialization (`adc_read()` does that lazily, but calling `adcStart()` from the S-Locked state is not allowed).

Reply:
BTW, your current code also has that ADC init issue; you may perform a dummy `adc_read()` in the init code to make sure that `adcStart()` has been called in the appropriate state.

Reply:
Having some interrupt delaying the `discharge_capacitor()` should really matter, I can definitely leave that one out especially since the `adc_read()` reenables interrupts as you said. I tested that and it works as intended. 
I'm inclined to use then the following and "leave it there" as tested working for now, until I come up with something more coherent in the actual ADC implementations.

```c
ATOMIC_BLOCK_FORCEON {
        // Set the row pin to high state and have capacitor charge
        charge_capacitor(row);
        // Read the ADC value
        sw_value = adc_read(adcMux);
    }
// Discharge peak hold capacitor
discharge_capacitor();
```

Regarding the dummy `adc_read()`, I was thinking about calling it in my `ecsm_init` function right after the `adcMux = pinToMux(ANALOG_PORT);`. 

If that's ok I'll commit these and push.

Reply:
<img width="633" alt="Screenshot 2023-02-23 at 5 12 28 PM" src="https://user-images.githubusercontent.com/40441626/220964928-a5af7504-024e-4c80-b5f9-c96cc05d460f.png">

<img width="497" alt="Screenshot 2023-02-23 at 5 12 46 PM" src="https://user-images.githubusercontent.com/40441626/220965020-47053440-1058-476c-8136-d1899f8a986a.png">